gPTP support for NXP i.MX NETC switch (no switch tag support)

yangbolu1991 opened the pull request:
gPTP had already been supported on NXP i.MX NETC switch via DSA tag driver on i.MX943. But for i.MX RT1180 NETC switch, there isn't switch tag support.

This PR is to support gPTP on NETC switch without switch tag. Then the switch PTP timestamping support is via ENETC host driver.

Also enabled RT1180 board in gPTP sample. gPTP worked well on all 4 switch ports and 1 end-point enetc port per my verification.

Thank you very much.

yangbolu1991 committed:
dts: arm: nxp_rt118x: add ptp clock node

Added ptp clock node.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
drivers: ptp_clock: support clock division for platforms

Supported clock division for platforms as it may vary from
platforms.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
drivers: dsa_nxp_imx_netc: improve PTP timestamping enablement

PTP timestamping should be enabled only when NET_L2_PTP and
PTP_CLOCK_NXP_NETC are selected. And NET_PKT_CONTROL_BLOCK
is required for PTP timestamping code.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
drivers: dsa_nxp_imx_netc: support PTP timestamping for no-tag switch

For NETC switch with tag support, PTP timestamping had already been
supported in tag driver.

This patch is to add PTP timestamping support for no-tag supported switch.
Timestamp functions are implemented and exported for host driver to
handle.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
drivers: eth_nxp_imx_netc: improve PTP timestamping enablement

Converted to use NETC_PTP_TIMESTAMPING_SUPPORT for PTP timestamping.
PTP timestamping is needed only when NET_L2_PTP and PTP_CLOCK_NXP_NETC
are selected.

Used default 16 bytes for ETH_NXP_IMX_RX_RING_LEN for PTP timestamping
because RX extended descriptor would be used.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
boards: nxp: mimxrt1180_evk: enable ptp clock

Enabled ptp clock.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
samples: net: gptp: add mimxrt1180_evk_mimxrt1189_cm33 support

Added mimxrt1180_evk_mimxrt1189_cm33 support.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
drivers: eth_nxp_imx_netc: support PTP timestamping for no-tag switch

Added PTP timestamping support in NETC ENETC host driver for no-tag
supported NETC switch.

On RX path, extended descriptor is used on host (pseudo MAC) to get
timestamp writeback.

On TX path, TX timestamp response should be enabled on host (pseudo MAC).
After TX, writeback TX timestamp ID should be updated to tx pkt fifo.
Once RX path received TX timestamp response, match TX timestamp to tx
pkt in fifo with the timestamp ID.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
dts: arm: nxp_rt118x: fix dts format

Fixed dts format.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
boards: nxp: mimxrt1180_evk: fix dtsi file format

Fixed dtsi file format.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 commented:
Appended two dts format fix-up patches to address CI compliance check.
Thanks.

jukkar reviewed (dismissed):
LGTM

yangbolu1991 commented:
Rebased to address conflict.
Thanks.

yangbolu1991 commented on samples/net/gptp/boards/mimxrt1180_evk_mimxrt1189_cm33.conf:
```diff
@@ -0,0 +1,7 @@
+CONFIG_NET_L2_ETHERNET=y
+CONFIG_ETH_DRIVER=y
```
Thanks @maass-hamburg 

Updated to v2. Changes include,
- Removed CONFIG_ETH_DRIVER in board conf which was default enabled.

maass-hamburg commented on drivers/ethernet/nxp_imx_netc/eth_nxp_imx_netc_psi.c:
```diff
@@ -216,8 +216,8 @@ static const struct ethernet_api netc_eth_api = {.iface_api.init = netc_eth_ifac
 			.rx_intr_msg_data = NETC_RX_INTR_MSG_DATA_START + n,))                     \
-		IF_ENABLED(CONFIG_PTP_CLOCK_NXP_NETC,				                   \
-			(.ptp_clock = DEVICE_DT_GET_OR_NULL(DT_INST_PHANDLE(n, ptp_clock)),))      \
+		IF_ENABLED(CONFIG_NET_L2_PTP, (IF_ENABLED(CONFIG_PTP_CLOCK_NXP_NETC,               \
```
```suggestion
		IF_ENABLED(NETC_PTP_TIMESTAMPING_SUPPORT,               \
```
why not like this, like the others?

yangbolu1991 replied:
Thanks. This is perfect.

Updated to v3. Changes include,
- Code improvement per @maass-hamburg suggestion.

Thanks.

maass-hamburg commented on soc/nxp/imxrt/imxrt118x/Kconfig.defconfig:
```diff
@@ -42,4 +42,7 @@ choice CACHE_TYPE
 	default EXTERNAL_CACHE if SOC_MIMXRT1189_CM33
 endchoice
 
+config PTP_CLOCK_NXP_NETC_CLK_DIV
```
use `configdefault`

yangbolu1991 commented on drivers/ptp_clock/Kconfig.nxp_netc:
```diff
@@ -8,3 +8,13 @@ config PTP_CLOCK_NXP_NETC
+
+config PTP_CLOCK_NXP_NETC_CLK_DIV
+	int
+	default 2
```
Thanks @maass-hamburg  It makes sense.
Per my understanding, imxrt1180 may be different having division 1, other platforms imx95/imx943 have division 2.
Let me use soc depended definition just in driver as there are just several line and no header file now.

yangbolu1991 commented:
Updated to v4. Changes include,
- Converted to use soc dependent definition for ptp clock division, instead of Kconfig.

@maass-hamburg could you help to review if it address your comments?
Thanks.
